### DISCUSSION TITLE
Query interrupt wait times

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#13868] [General]
I frequently run into the situation that I accidentally start a query that I quickly realize does too much, then try to send a KeyboardInterrupt and this interrupt taking more than >30s, without the machine being at max memory usage. 

This is rather inconvenient in interactive workflows. if you agree that duckdb should be able to handle interrupts near instantly (from experience pandas/polars/numpy tend to better at this), I am happy to collect some reproducing examples.

**Comment by Tishj** [comment]
Reproductions would be great, then these can be investigated / addressed 👍 
It depends on the complexity of the operation, at least a chunk of tuples will have to be processed before an interrupt gets honored

**Comment by soerenwolfers** [comment]
@Tishj Here is one: If I run the snippet below in a Jupyter notebook and then interrupt after a few seconds, then about 60% of the time it interrupts instantly and 40% of the time I have to wait 30 seconds for the interrupt to be actioned. If I run and interrupt the same snippet in the CLI it interrupts instantly always.

[interrupts.webm](https://github.com/user-attachments/assets/631fdb08-168a-49e6-9a0c-92040794db37)



```python
import duckdb
duckdb.query("""
COPY (FROM range(1_000_000_000)) TO '/tmp/test.parquet'
""")
```

**Reply by Tishj** [reply]
I'm not sure, but it might be the same issue that <https://github.com/duckdb/duckdb/pull/14034> addresses?

**Reply by Tishj** [reply]
This was discovered in `pg_duckdb` but the way interruptions are slightly delayed is identical to python, where we are at the mercy of ExecuteTask returning before we can check for an interrupt exception

```c++
unique_ptr<QueryResult> DuckDBPyConnection::CompletePendingQuery(PendingQueryResult &pending_query) {
	PendingExecutionResult execution_result;
	while (!PendingQueryResult::IsResultReady(execution_result = pending_query.ExecuteTask())) {
		{
			py::gil_scoped_acquire gil;
			if (PyErr_CheckSignals() != 0) {
				throw std::runtime_error("Query interrupted");
			}
		}
		if (execution_result == PendingExecutionResult::BLOCKED) {
			pending_query.WaitForTask();
		}
	}
```

**Reply by soerenwolfers** [reply]
Would it work to
1. at the beginning of `duckdb.query`, save the current Python level signal handlers using `previous = signal.getsignal(signal.SIGINT)` and then in C++ code register a new handler using `std::signal` in the same way as the CLI
2. execute the query in Python-unaware C++
3. at the end of `duckdb.query` restore the Python handlers `signal.signal(signal.SIGINT, previous)`, and if the return value of step 2 indicates that it handled a SIGINT, call those Python handlers

?

**Reply by Tishj** [reply]
Perhaps, but I'm not super comfortable with messing with pythons signal handlers

Might be worth looking into

**Reply by Tishj** [reply]
The problem is also that we call into Python, if we do that without their signal handler and sigint is send I could see everything crashing

**Reply by soerenwolfers** [reply]
But you don't call Python from within the "slow" code, right? My idea would have been (in pseudo-code using Python syntax, but I imagine this is all being in the C++ wrapper code

```python
<preliminary steps that don't take long>
PyErr_CheckSignals()
previous_handler = get_signal_handler(SIGINT)
try:
    set_signal_handler(SIGINT, whatever_signal_handler_the_CLI_installs)
    
    # execute_query is duckdb C++ library code that doesn't know anything about python;
    # it periodically checks for a flag set by the signal handler 
    # and returns prematurely if the flag is set, encoding this fact in its output.
    result = execute_query()

    if result.error_or_sigint:
        raise Exception()
finally:
    set_signal_handler(SIGINT, previous_handler)
```

To make this extra safe, maybe only swap out the signal handler if duckdb is running on the main thread and duckdb holds the GIL?

In any case, if there is an alternative you're more comfortable with, that's probably better. I'm far from knowledgeable about this topic myself.

**Reply by Tishj** [reply]
That sounds great but isn't the reality unfortunately
There are many paths in the execution that could call into python, one example is scanning from pandas

**Reply by Tishj** [reply]
Which means we would have to set and reset the signal handler every time we grab the gil
